### PR TITLE
drivers: can: add support for nxp can fd flexcan

### PR DESCRIPTION
**Pull request by dipakgmx**
`drivers: can:`  Config flag CAN_MCUX_FLEXCAN_FD added to enable can fd on nxp, flexcan-fd compatible node. New binding for nxp flexcan fd added. New driver module for flexcan-fd can_mcux_flexcan_fd is introduced. This is distinct from the previous flexcan module which does not support can fd. Due to a significant overlap of the code between the 2 modules, the common code sections between them is now moved to a new module called
 can_mcux_flexcan_common.
`dts: bindings: can:` New binding for nxp flexcan-fd introduced
`dts: arm: nxp:` Changed flexcan compatibility to flexcan-fd on nxp_rt1060.dtsi & nxp_rt11xx.dtsi
`soc: arm : nxp_imx: rt:` added additional compiler guard CONFIG_CAN_MCUX_FLEXCAN_FD for flexcan
`drivers: clock_control:` added additional compiler guard CONFIG_CAN_MCUX_FLEXCAN_FD for flexcan 
`boards: arm: mimxrt1060_evk:` The flexcan3 node is updated to flexcan-fd due to the introduction of the new flexcan-fd compatible. Previously used flexcan2 node is now disabled and the flexcan3 node replaces it as the chosen node. Pins formerly used by the flexcan2 node is now used by flexcan3.
`boards: arm: mimxrt1160_evk:` flexcan-fd support added 
`boards: arm: mimxrt1170_evk:` flexcan-fd support added

### CODE REVIEW

**Comment by dipakgmx** [comment]
I was able to test this on my NXP MIMXRT1060-EVK board using a [modified ](https://github.com/dipakgmx/zephyr/tree/feature_mimxrt1060_can_fd_counter_sample)loopback test for CAN FD. Could someone test it on their end too?

File: dts/bindings/can/nxp,canfd-flexcan.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# Copyright (c) 2022 Carl Zeiss Meditec AG
+# SPDX-License-Identifier: Apache-2.0
+
+description: NXP CANFD FlexCAN controller
+
+compatible: "nxp,canfd-flexcan"
```
Comment (henrikbrixandersen) [inline dts/bindings/can/nxp,canfd-flexcan.yaml:6]:
```suggestion
compatible: "nxp,flexcan-fd"
```

I would suggest using `nxp,flexcan-fd` to better signal the relationship with the classic `nxp,flexcan` compatible. @dleach02 et al. Any preference here? The old one has the `nxp,kinetis-flexcan` compatible, but perhaps it ought to be renamed to just `nxp,flexcan`?

File: dts/arm/nxp/nxp_rt1060.dtsi [hunk 2]
```diff
@@ -27,7 +27,9 @@
 	};
 };
 
-/* i.MX rt1060 has a second Ethernet controller. */
+/* i.MX rt1060 has a second Ethernet controller.
+ * The FlexCAN3 is capble of CAN FD.
+ */
```
Comment (henrikbrixandersen) [inline dts/arm/nxp/nxp_rt1060.dtsi:32]: These comments seem unrelated. I see no reason for bundling them into one.

File: dts/arm/nxp/nxp_rt1060.dtsi [hunk 3]
```diff
@@ -43,6 +45,13 @@
 				interrupt-names = "IEEE1588_TMR";
 			};
 		};
+
+		flexcan3: can@401d8000 {
+			compatible = "nxp,canfd-flexcan";
+			bus-speed-data = <2000000>;
```
Comment (henrikbrixandersen) [inline dts/arm/nxp/nxp_rt1060.dtsi:51]: `bus-speed` should not be set on the SoC level. Also, I believe you're missing `sjw` and `sample-point` properties?

File: drivers/can/can_mcux_flexcan.c [hunk 4]
```diff
@@ -4,7 +4,11 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#ifdef CONFIG_DT_HAS_NXP_CANFD_FLEXCAN_ENABLED
+#define DT_DRV_COMPAT nxp_canfd_flexcan
+#else
 #define DT_DRV_COMPAT nxp_kinetis_flexcan
+#endif
```
Comment (henrikbrixandersen) [inline drivers/can/can_mcux_flexcan.c:11]: This won't work. These SoCs have both classic FlexCAN controllers and FlexCAN-FD controllers. We need to be able to use both at the same time. Your changes to the classic FlexCAN driver only ever allows using one type or the other.

Given that this is a completely different hardware IP (which the NXP MCUX HAL wraps in a similar API as the classic FlexCAN) - and the amount of changes to the existing, classic FlexCAN driver - I think I would go for a completely new driver (with core FlexCAN Zephyr shim functions split into a common C file).

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: dts/bindings/can/nxp,canfd-flexcan.yaml (hunk 1 above)
Comment (dleach02) [reply dts/bindings/can/nxp,canfd-flexcan.yaml:6]: let me discuss internally on this point

File: dts/arm/nxp/nxp_rt1060.dtsi (hunk 2 above)
Comment (dipakgmx) [reply dts/arm/nxp/nxp_rt1060.dtsi]: Fixed this with the latest commit.

File: drivers/can/can_mcux_flexcan.c (hunk 4 above)
Comment (dipakgmx) [reply drivers/can/can_mcux_flexcan.c]: For now, I have split them into 2 different files. One pertaining to the classic FlexCAN and the other for the FlexCAN-FD. Let me know if this is what we are looking for. I'll refactor the common functions across the files in the next commit.

File: dts/bindings/can/nxp,canfd-flexcan.yaml (hunk 1 above)
Comment (dleach02) [reply dts/bindings/can/nxp,canfd-flexcan.yaml:6]: Agree with using "flexcan-fs"

Comment (dipakgmx) [reply dts/bindings/can/nxp,canfd-flexcan.yaml]: Fixed now.

File: dts/arm/nxp/nxp_rt1060.dtsi (hunk 3 above)
Comment (dipakgmx) [reply dts/arm/nxp/nxp_rt1060.dtsi]: Removed the `bus_speed`.

File: drivers/can/can_mcux_flexcan.c (hunk 4 above)
Comment (dipakgmx) [reply drivers/can/can_mcux_flexcan.c]: Splitting the common functions doesn't seem feasible, since the types in the functions are more or less different.

File: soc/arm/nxp_imx/rt/soc_rt10xx.c [hunk 5]
```diff
@@ -224,7 +224,7 @@ static ALWAYS_INLINE void clock_init(void)
 	CLOCK_SetDiv(kCLOCK_CsiDiv, 0); /* Set CSI divider to 1 */
 	CLOCK_SetMux(kCLOCK_CsiMux, 0); /* Set CSI source to OSC 24M */
 #endif
-#ifdef CONFIG_CAN_MCUX_FLEXCAN
+#if defined(CONFIG_CAN_MCUX_FLEXCAN) || (CONFIG_CAN_MCUX_FLEXCAN_FD)
```
Comment (henrikbrixandersen) [inline soc/arm/nxp_imx/rt/soc_rt10xx.c:227]:
```suggestion
#if defined(CONFIG_CAN_MCUX_FLEXCAN) || defined(CONFIG_CAN_MCUX_FLEXCAN_FD)
```

File: soc/arm/nxp_imx/rt/soc_rt11xx.c [hunk 6]
```diff
@@ -407,7 +407,7 @@ static ALWAYS_INLINE void clock_init(void)
 	CLOCK_SetRootClock(kCLOCK_Root_Lpspi1, &rootCfg);
 #endif
 
-#ifdef CONFIG_CAN_MCUX_FLEXCAN
+#if defined(CONFIG_CAN_MCUX_FLEXCAN) || (CONFIG_CAN_MCUX_FLEXCAN_FD)
```
Comment (henrikbrixandersen) [inline soc/arm/nxp_imx/rt/soc_rt11xx.c:410]:
```suggestion
#if defined(CONFIG_CAN_MCUX_FLEXCAN) || defined(CONFIG_CAN_MCUX_FLEXCAN_FD)
```

File: drivers/can/can_mcux_flexcan_fd.c [hunk 7]
```diff
@@ -0,0 +1,1091 @@
...
+	uint32_t mcr;
+	uint32_t fdctrl;
+
+	if (data->started) {
+		return -EBUSY;
+	}
+
+
+	if ((mode &
+	     ~(CAN_MODE_FD | CAN_MODE_LOOPBACK | CAN_MODE_LISTENONLY | CAN_MODE_3_SAMPLES)) != 0) {
```
Comment (henrikbrixandersen) [inline drivers/can/can_mcux_flexcan_fd.c:307]: Where do you handle the `CAN_MODE_FD` flag? To me it looks like this driver does not handle switching between classic and CAN-FD mode?

File: drivers/can/can_mcux_flexcan_fd.c [hunk 8]
```diff
@@ -0,0 +1,1091 @@
...
+		dest->flags |= CAN_FRAME_IDE;
+		dest->id = FLEXCAN_ID_TO_CAN_ID_EXT(src->id);
+	}
+
+	if (src->type == kFLEXCAN_FrameTypeRemote) {
+		dest->flags |= CAN_FRAME_RTR;
+	}
+
+	dest->dlc = src->length;
+	dest->flags |= CAN_FRAME_FDF;
```
Comment (henrikbrixandersen) [inline drivers/can/can_mcux_flexcan_fd.c:398]: You cannot unconditionally set the `CAN_FRAME_FDF` bit. The driver should work in both classic and CAN-FD modes.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
- Please split your commit into several commits, each dealing with an isolated change (e.g. one for adding the dts binding, one for adding the driver, one for adding SoC support, one for adding clock control support and so on). Breaking the changes down into logically separate changes greatly improves the review experience and the Git traceability/history.
- It looks like you have written a CAN-FD-only driver? Does this driver pass the tests tagged with "can" (`twister --device-testing ... -t can`)? Even the non-CAN-FD tests?
- Please add an EVK board configuration enabling this driver, otherwise it will just sit there and never be compile tested/automatically tested in neither CI nor on test board farms.
- I am still not convinced about the amount of duplicated code between the existing FlexCAN driver and this new FlexCAN-FD driver. Looks like this could be greatly reduced.
- You add SoC support for both rt10xx and rt11xx, but yet only add dts support to rt1060?

File: soc/arm/nxp_imx/rt/soc_rt10xx.c (hunk 5 above)
Comment (dipakgmx) [reply soc/arm/nxp_imx/rt/soc_rt10xx.c]: Resolved.

File: soc/arm/nxp_imx/rt/soc_rt11xx.c (hunk 6 above)
Comment (dipakgmx) [reply soc/arm/nxp_imx/rt/soc_rt11xx.c]: Resolved.

File: drivers/clock_control/clock_control_mcux_ccm_rev2.c [hunk 9]
```diff
@@ -77,7 +77,7 @@ static int mcux_ccm_get_subsys_rate(const struct device *dev,
 		break;
 #endif
 
-#ifdef CONFIG_CAN_MCUX_FLEXCAN
+#if defined(CONFIG_CAN_MCUX_FLEXCAN) || (CONFIG_CAN_MCUX_FLEXCAN_FD)
```
[the remark this answers is not in the document]
Comment (dipakgmx) [reply drivers/clock_control/clock_control_mcux_ccm_rev2.c]: Resolved.

File: drivers/clock_control/clock_control_mcux_ccm.c [hunk 10]
```diff
@@ -143,7 +143,7 @@ static int mcux_ccm_get_subsys_rate(const struct device *dev,
 	} break;
 #endif
 
-#ifdef CONFIG_CAN_MCUX_FLEXCAN
+#if defined(CONFIG_CAN_MCUX_FLEXCAN) || (CONFIG_CAN_MCUX_FLEXCAN_FD)
```
[the remark this answers is not in the document]
Comment (dipakgmx) [reply drivers/clock_control/clock_control_mcux_ccm.c]: Resolved.

File: drivers/can/Kconfig.mcux [hunk 11]
```diff
@@ -12,12 +12,41 @@ config CAN_MCUX_FLEXCAN
...
+	default 3
+	help
+	  CAN FD FlexCAN Message Buffer payload size. The actual
+	  transmitted or received CAN FD frame data size needs
+	  to be less than or equal to this value.
+	  The following options are available:
+	  <0>: 8 bytes per message buffer.
+	  <1>: 16 bytes per message buffer.
+	  <2>: 32 bytes per message buffer.
+	  <3>: 64 bytes per message buffer.
```
[the remark this answers is not in the document]
Comment (dipakgmx) [reply drivers/can/Kconfig.mcux]: Eliminated this additional flag. Now when CAN-FD is activated, the message size is defaulted to 64 bytes.

File: drivers/can/can_mcux_flexcan_fd.c (hunk 7 above)
Comment (dipakgmx) [reply drivers/can/can_mcux_flexcan_fd.c:197]: Fixed this issue now. [Here](https://github.com/dipakgmx/zephyr/blob/feature_mimxrt1060_can_fd_support/drivers/can/can_mcux_flexcan_fd.c#L238) I allow transmission of CAN 2.0 frames.

File: drivers/can/can_mcux_flexcan_fd.c (hunk 8 above)
Comment (dipakgmx) [reply drivers/can/can_mcux_flexcan_fd.c]: [Now](https://github.com/dipakgmx/zephyr/blob/feature_mimxrt1060_can_fd_support/drivers/can/can_mcux_flexcan_fd.c#L270) the received flags are set and the can frame type flags are set correspondingly.

**Comment by dipakgmx** [comment]
> * Please split your commit into several commits, each dealing with an isolated change (e.g. one for adding the dts binding, one for adding the driver, one for adding SoC support, one for adding clock control support and so on). Breaking the changes down into logically separate changes greatly improves the review experience and the Git traceability/history.

Split the commits now based on the context.

> * It looks like you have written a CAN-FD-only driver? Does this driver pass the tests tagged with "can" (`twister --device-testing ... -t can`)? Even the non-CAN-FD tests?
> * Please add an EVK board configuration enabling this driver, otherwise it will just sit there and never be compile tested/automatically tested in neither CI nor on test board farms.

Fixed this. I tested them on my boards now. 
Added the new board config, and running `python scripts/twister --device-testing --device-serial COM3 --device-serial-baud 115200 -p mimxrt1060_evk_canfd -n -T tests/drivers/can` succeeds.

> * I am still not convinced about the amount of duplicated code between the existing FlexCAN driver and this new FlexCAN-FD driver. Looks like this could be greatly reduced.

Eliminated clones across the code. Reused code is now called in a new `common` module.

> * You add SoC support for both rt10xx and rt11xx, but yet only add dts support to rt1060?
Fixed this now.

File: drivers/can/can_mcux_flexcan_fd.c (hunk 7 above)
Comment (dipakgmx) [reply drivers/can/can_mcux_flexcan_fd.c:225]: Fixed this now. Am able to use the driver in classic CAN mode.

**Comment by dipakgmx** [comment]
Hi @henrikbrixandersen 

I tried fixing the code clone issues using a quasi-template approach in C. These changes are in [another branch](https://github.com/dipakgmx/zephyr/tree/feature_mimxrt1060_can_fd_support_flexcan_templated_driver).

The general idea behind this is having 2 sets of types/function, one for the `classic` case and the other for `fd`. Within the file `can_mcux_flexcan_template`, the functions are declared/defined and the relevant code is generated via the preprocessor flag `CAN_TYPE__`. The flag `CAN_TYPE__` would be set in the relevant compilation unit, say either `can_mcux_flexcan_fd` or `can_mcux_flexcan.c` and this would generate the relevant functions. 

With this, am able to step through the code when debugging. Also errors during compilation are shown as the relevant lines in the template source file `can_mcux_flexcan_template.c`.

Am uncertain if this could be a solution. Could you let me know your opinion about it?

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
There is no need for defining a set of completely separate boards in order to use CAN-FD on the NXP mimxrt10xx. The `compatible` for FlexCAN3 just needs to be updated on the SoC level to indicate, that it supports CAN-FD.

I must admit that I do not like the whole template approach. It makes the code much more difficult to read, debug etc.

**Comment by dipakgmx** [comment]
> There is no need for defining a set of completely separate boards in order to use CAN-FD on the NXP mimxrt10xx. The `compatible` for FlexCAN3 just needs to be updated on the SoC level to indicate, that it supports CAN-FD.

I think, I have [added](https://github.com/zephyrproject-rtos/zephyr/pull/52068/files#diff-425b678c92afec4bb4f3fd999d21ac28625333fdff0c1be4a368f3bcd609938d) it under the SoC level. Regarding the later, do you mean to suggest swapping the `&flexcan2` with `flexcan3` [here](https://github.com/dipakgmx/zephyr/blob/feature_mimxrt1060_can_fd_support/boards/arm/mimxrt1060_evk/mimxrt1060_evk.dts#L33)?

**Comment by henrikbrixandersen** [comment]
> I think, I have [added](https://github.com/zephyrproject-rtos/zephyr/pull/52068/files#diff-425b678c92afec4bb4f3fd999d21ac28625333fdff0c1be4a368f3bcd609938d) it under the SoC level. Regarding the later, do you mean to suggest swapping the `&flexcan2` with `flexcan3` [here](https://github.com/dipakgmx/zephyr/blob/feature_mimxrt1060_can_fd_support/boards/arm/mimxrt1060_evk/mimxrt1060_evk.dts#L33)?

Yes.

**Comment by dipakgmx** [comment]
> Adding another board is problematic because we already have an mimxrtRT1060_evk and a mimxrt1060_evkb. This PR creates a new board based on the rev A board and not rev B.

I have eliminated the newly created boards and reverted back to to the old boards. I have have replace the default can node from `flexcan2` to `flexcan3` since it supports CAN FD. But I have retained the `flexcan2` node in the mimxrtRT1064_evk board so that we can test both CAN 2.0 and CAN FD in the build pipelines. Would this make sense?

**Comment by dipakgmx** [comment]
> I must admit that I do not like the whole template approach. It makes the code much more difficult to read, debug etc.

I have reverted back to the old approach now, with 2 distinct files. I have refactored a majority of the possible functions into the common module. The twister hardware tests pass with the reconfigured CAN FD compatible mimxrtRT1060_evk board. Running the same tests when in reconfigure the board back to CAN 2.0 (with `flexcan2`) I encounter problems with building can.shell tests. It just doenst seem to build.

**Comment by bbuzzihg** [comment]
Hi Dipak, I'm also currently working on the problem to get CAN-FD with my RT1060 EVKB working and was happy to find your PR here. Thanks for your work! 
Based on your changes I was able to get my application to build and the CAN-FD stack is successfully initialized and started. 

Sending frames to the CAN transceiver seems to work, however, I'm not able to get any signal on my CAN bus which is connected via pins gpio_ad_b0_14, gpio_ad_b0_15.
You already reconfigured the pinmux_flexcan3 to use those pins and I also tried to directly change it in mimxrt1060_evk-pinctrl.dtsi, but no success. Non-FD CAN via flexcan2 is working without a problem from those pins.

Do you still have similar problems or can you help me which pieces am I missing here to get a first test implementation running?
I'm pretty new to Zephyr, but I'm happy to contribute here with my tests or any support I can give.

**Comment by dipakgmx** [comment]
Hi [bbuzzihg](https://github.com/bbuzzihg),

Thanks for the feedback.

> You already reconfigured the pinmux_flexcan3 to use those pins and I also tried to directly change it in mimxrt1060_evk-pinctrl.dtsi, but no success. Non-FD CAN via flexcan2 is working without a problem from those pins.

I think I have made a mistake here. Could you try changing the `mimxrt1060_evk-pinctrl.dtsi` pins like this:

```
/* Option 1 */
&pinctrl {
pinmux_flexcan3: pinmux_flexcan3 {
		group0 {
			pinmux = <&iomuxc_gpio_ad_b0_14_flexcan3_tx>,
			<&iomuxc_gpio_ad_b0_15_flexcan3_rx>;
		};
	};
};
```

**Comment by bbuzzihg** [comment]
Hey @dipakgmx, thanks a lot, I just did a retest with my board and CAN-FD RX and TX now work perfectly fine 👍 
The only thing I needed to change was to lower the bus-speed-data to 1000KBit. With 5000KBit setting the data bit rate (can_set_bitrate_data) during initialization fails, at least for my RT 1060 EVKB board.
I will continue with more tests during the next days.

**Comment by dipakgmx** [comment]
> With 5000KBit setting the data bit rate (can_set_bitrate_data) during initialization fails with error code -134 (ENOTSUP), at least for my RT 1060 EVKB board.

Hi @bbuzzihg, was this test done with the new commits I pushed? I will grab a EVKB board probably tomorrow and test the same.

**Comment by dipakgmx** [comment]
> Hey @dipakgmx, thanks a lot, I just did a retest with my board and CAN-FD RX and TX now work perfectly fine 👍 The only thing I needed to change was to lower the bus-speed-data to 1000KBit. With 5000KBit setting the data bit rate (can_set_bitrate_data) during initialization fails with error code -134 (ENOTSUP), at least for my RT 1060 EVKB board. I will continue with more tests during the next days.

I don't seem to get any errors when running on the RT 1060 EVK board. The boards are very similar except for the green LED. Could you do the following:

1. See if this test functions correctly? This sends out a incremental counter value.

```
#include <stdio.h>

#include <zephyr/device.h>
#include <zephyr/drivers/can.h>
#include <zephyr/kernel.h>
#include <zephyr/logging/log.h>
#include <zephyr/logging/log_ctrl.h>
#include <zephyr/sys/byteorder.h>
#include <zephyr/sys/printk.h>

#define COUNTER_MSG_ID 0x12345
#define SLEEP_TIME K_MSEC(250)
LOG_MODULE_REGISTER(main, CONFIG_CAN_LOG_LEVEL);

const struct device *const can_dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_canbus));

void main(void)
{
	struct can_frame counter_frame = {
#ifdef CONFIG_CAN_FD_MODE
                .flags = CAN_FRAME_IDE | CAN_FRAME_FDF | CAN_FRAME_BRS,
#else
                .flags = CAN_FRAME_IDE,
#endif
		.id = COUNTER_MSG_ID,
		.dlc = 2
	};
	uint32_t counter = 0;
	int ret;

	if (!device_is_ready(can_dev)) {
                LOG_PRINTK("CAN: Device %s not ready.\n", can_dev->name);
		return;
	}

	ret = can_start(can_dev);
	if (ret != 0) {
                LOG_PRINTK("Error starting CAN controller [%d]", ret);
		return;
	}
        struct can_timing timing;
        struct can_timing timing_data;

        ret = can_calc_timing(can_dev, &timing, 125000, 875);
        if (ret > 0) {
                LOG_INF("Sample-Point error: %d", ret);
        }

        if (ret < 0) {
                LOG_ERR("Failed to calc a valid timing");
                return;
        }

#ifdef CONFIG_CAN_FD_MODE
        ret = can_calc_timing_data(can_dev, &timing_data, 5000000, 875);
        if (ret >= 0) {
                LOG_INF("Sample-Point data error: %d", ret);
        }

        if (ret < 0) {
                LOG_ERR("Failed to calc a valid timing data");
                return;
        }
#endif

        ret = can_stop(can_dev);
        if (ret != 0) {
                LOG_ERR("Failed to stop CAN controller");
        }

        ret = can_set_timing(can_dev, &timing);
        if (ret != 0) {
                LOG_ERR("Failed to set timing");
        }

#ifdef CONFIG_CAN_FD_MODE
        ret = can_set_timing_data(can_dev, &timing_data);
        if (ret != 0) {
                LOG_ERR("Failed to set timing data");
        }
#endif

        ret = can_start(can_dev);
        if (ret != 0) {
                LOG_ERR("Failed to start CAN controller");
        }

        LOG_INF("Finished init.\n");

	while (1) {

                UNALIGNED_PUT(sys_cpu_to_be16(counter),
                              (uint16_t *)&counter_frame.data[0]);
                counter++;
                /* This sending call is blocking until the message is sent. */
                int status = can_send(can_dev, &counter_frame, K_MSEC(100), NULL, NULL);
                if (status != 0)
                {
                  LOG_PRINTK("Frame transmission error: %d\n", status);
                }
                LOG_PRINTK("Counter: %u\n", counter);
		k_sleep(SLEEP_TIME);
	}
}
```

2. Run twister tests on your board: (pick the serial port for your machine, in my case it was `COM3`.

```
python scripts/twister --short-build-path --device-testing --device-serial COM3 --device-serial-baud 115200 -p mimxrt1060_evkb  -T tests/drivers/can/
```

**Comment by henrikbrixandersen** [comment]
Please see #56419 for an alternative proposal.
